R/forecastingdata_org.R

```diff
@@ -0,0 +1,380 @@
+  
+  while(length(line) && regexpr('^[[:space:]]*@data', line, perl = TRUE) == -1) { # Until read @data, run this loop (-1 indicate no match with the regular expression yet)
+    
+    if(regexpr('^[[:space:]]*@', line, perl = TRUE) > 0) { # This condition will be true for lines starting with @
+      
+      con <- textConnection(line)
+      line <- scan(con, character(), quiet = TRUE) # Creating a vector containing the words in a line (ex: "@attribute" "series_name" "string")
+      close(con)
+      
+      if(line[1] == "@attribute"){
+        if(length(line) != 3)  # Attributes have both name and type
+          stop("Invalid meta-data specification.")
+        
+        if(is.null(index) & line[3] == "date") # Searching for a valid index, if index is not given 
+          index_var <- line[2]
+        
+        col_names <- c(col_names, line[2])
+        col_types <- c(col_types, line[3])
+      }else{
+        if(length(line) != 2) # Other meta-data have only values
+          stop("Invalid meta-data specification.")
+        
+        if(line[1] == "@frequency")
+          frequency <- line[2]
+        else if(line[1] == "@horizon")
+          forecast_horizon <- as.numeric(line[2])
+        else if(line[1] == "@missing")
+          contain_missing_values <- as.logical(line[2])
+        else if(line[1] == "@equallength")
+          contain_equal_length <- as.logical(line[2])
+      }
+    }
+    line <- readLines(file, n = 1)
+  }
+  
+  if(length(line) == 0)
+    stop("Missing data section.")
+  if(is.null(col_names))
+    stop("Missing attribute section.")
+  
+  line <- readLines(file, n = 1)
+  
+  if(length(line) == 0)
+    stop("Missing series information under data section.")
+  
+  for(col in col_names)
+    assign(col, NULL)
+  
+  values <- NULL
+  row_count <- 0
+  
+  # Get data
+  while(length(line) != 0){
+    full_info <- strsplit(line, ":")[[1]]
+    
+    if(length(full_info) != length(col_names)+1)
+      stop("Missing attributes/values in series.")
+    
+    series <- strsplit(tail(full_info, 1), ",")[[1]]
+    series[which(series == "?")] <- NA
+    series <- as.numeric(series)
+    
+    if(sum(is.na(series)) == length(series))
+      stop("All series values are missing. A given series should contains a set of comma separated numeric values. At least one numeric value should be there in a series.")
+    
+    values <- c(values, series)
+    row_count <- row_count + length(series)
+    
+    attributes <- head(full_info, length(full_info)-1)
+    
+    for(col in seq_along(col_names)){
+      
+      att <- eval(parse(text=col_names[col]))
+      
+      # This format supports 3 attribute types: string, numeric and date
+      if(col_types[col] == "date"){
+        if(is.null(frequency))
+          stop("Frequency is missing.")
+        else{
+          if(frequency %in% LOW_FREQUENCIES)
+            start_time <- as.POSIXct(attributes[col], format = "%Y-%m-%d %H-%M-%S")
```

Comment:
It could be simpler to first read the data in its original format, and then later drop precision to a Date class if appropriate for the frequency.